astropy/io/fits/diff.py

```diff
@@ -1240,6 +1242,19 @@ def report_diff_values(fileobj, a, b, ind=0):
             line = 'b>' + line[1:]
         else:
             line = ' ' + line
+        
+        if typea != typeb:
+            if 'a>' in line:
+                line += ' ' + repr(typea)
+            elif 'b' in line:
+                line += ' ' + repr(typeb)
+            elif '?' not in line:
+                line1 = 'a>' + line + ' ' + repr(typea)
+                fileobj.write(indent(u('  %s\n') % line1.rstrip('\n'), ind))
+                line2 = 'b>' + line + ' ' + repr(typeb)
+                fileobj.write(indent(u('  %s\n') % line2.rstrip('\n'), ind))
+                return
```

Comment:
I wouldn't really go about this by trying to parse things out of the diff output.

The main source of confusion in the original issue is numbers and strings that represent that same number.  So what I would do is actually make an explicit check early on to see if one of the values is a string and if one isn't e.g. 

```python
if (isinstance(a, string_types) and not isinstance(b, string_types)
        isinstance(b, string_types) and not isinstance(a string_types)):
```
If they are different then take the one that *is* a string (say, `a`) and do `a = repr(a).lstrip('u')` so that what we end up comparing to b is the value of `a` represented with quotes around it, distinguishing it as a string (the `lstrip('u')` is for compatibility with Python 2 unicode strings).  Then it will be pretty guaranteed to be different from whatever `b` is.

I do like that you included the types in the diff output though.  But I'd put it at the begging of each line like:

```
(float) a> 1.0
  (str) b> '1.0'
```
you can do this with `typea.__name__`, etc.  You'll also need to do some right-justification so that things line up with the types of different-length names (as in the above example).  Take a look at some of Python's advanced string formatting features, which have this capability built-in.